numpy/core/tests/test_dtype.py

```diff
@@ -446,6 +447,149 @@ def test_alignment(self):
         assert_equal(t1.alignment, t2.alignment)
 
 
+@pytest.mark.skipif(not HAS_REFCOUNT, reason="Python lacks refcounts")
+class TestStructuredObjectRefcounting:
+    """These tests cover various uses of complicated structured types which
+    include objects and thus require reference counting.
+    """
+    def iter_struct_object_dtypes(self, obj=None):
+        """
+        Iterates over a few complex dtypes and object pattern which
+        fill the array with a given object (defaults to a singleton).
+
+        Parameters
+        ----------
+        obj : object or None
+            The returned pattern can be used create an array with this objects
+            at all positions (which store python objects). If None, will
+            create a new object().
+
+        Yields
+        ------
+        dtype : dtype
+        pattern : tuple
+            Structured tuple for use with `np.array`.
+        count : int
+            Number of objects stored in the dtype.
+        """
+        if obj is None:
+            obj = object()
+
+        dt = np.dtype([('b', 'O', (2, 3))])
+        p = ([[obj] * 3] * 2,)
+        yield dt, p, 6
+
+        dt = np.dtype([('a', 'i4'), ('b', 'O', (2, 3))])
+        p = (0, [[obj] * 3] * 2)
+        yield dt, p, 6
+
+        dt = np.dtype([('a', 'i4'),
+                       ('b', [('ba', 'O'), ('bb', 'i1')], (2, 3))])
+        p = (0, [[(obj, 0)] * 3] * 2)
+        yield dt, p, 6
+
+        dt = np.dtype([('a', 'i4'),
+                       ('b', [('ba', 'O'), ('bb', 'O')], (2, 3))])
+        p = (0, [[(obj, obj)] * 3] * 2)
+        yield dt, p, 12
+
+    def test_structured_object_create_delete(self):
+        """Structured object reference counting in creation and deletion"""
+        # Create dummy objects, defining `one` and `zero` seemed to
+        # stabilize their reference count.
+        # The test assumes that 0, 1, and None are singletons.
+        singleton = object()
+        one = 1
+        zero = 0
+        for dt, pat, count in self.iter_struct_object_dtypes(singleton):
```

Comment:
It would be great to use `pytest.mark.paramtrize` here, rather than putting the for loop in the test. you might want to add a fourth `label` item to the tuple, so that you can do:
```
@pytest.mark.parametrize(['dt', 'pat', 'count', 'label'], iter_struct_object_dtypes(), id=lambda t: t[3])
def test_structured_object_create_delete(dt, pat, count, label):
    # don't actually use label here, pytest will have already handled it.
```